### PR TITLE
add hugetlbfs and hugepages handling for containers

### PR DESCRIPTION
**Pull request by brauner**
Signed-off-by: Christian Brauner <christian.brauner@ubuntu.com>

### CODE REVIEW

File: doc/api-extensions.md [hunk 1]
```diff
@@ -926,3 +926,9 @@ Introduces the ability to create a storage pool from an existing non-empty volum
 This option should be used with care, as LXD can then not guarantee that volume name conflicts won't occur
 with non-LXD created volumes in the same volume group.
 This could also potentially lead to LXD deleting a non-LXD volume should name conflicts occur.
+
+## container\_syscall\_intercept\_hugetlbfs
+When mount syscall interception is enabled and hugetlbfs is specified as an
+allowed filesystem type LXD will mount a separate hugetlbfs instance for the
+container with the uid and gid mount options set to the container's root uid
+and gid. This ensure that processes in the container can use hugepages.
```
Comment (tomponline) [inline doc/api-extensions.md]: @brauner just a nit "ensures",

**Comment by brauner** [comment]
On Mon, Mar 02, 2020 at 08:41:10AM -0800, TomP wrote:
> tomponline commented on this pull request.
> 
> 
> 
> > @@ -926,3 +926,9 @@ Introduces the ability to create a storage pool from an existing non-empty volum
>  This option should be used with care, as LXD can then not guarantee that volume name conflicts won't occur
>  with non-LXD created volumes in the same volume group.
>  This could also potentially lead to LXD deleting a non-LXD volume should name conflicts occur.
> +
> +## container\_syscall\_intercept\_hugetlbfs
> +When mount syscall interception is enabled and hugetlbfs is specified as an
> +allowed filesystem type LXD will mount a separate hugetlbfs instance for the
> +container with the uid and gid mount options set to the container's root uid
> +and gid. This ensure that processes in the container can use hugepages.
> 
> @brauner just a nit "ensures",

Woops, thanks!

File: lxd/seccomp/seccomp.go [hunk 2]
```diff
@@ -1249,6 +1249,15 @@ func mountFlagsToOpts(flags C.ulong) string {
 	return opts
 }
 
+// mountHandleHugetlbfsArgs adds user namespace root uid and gid to the
+// hugetlbfs mount options to make it useable in unprivileged containers.
+func (s *Server) mountHandleHugetlbfsArgs(args *MountArgs, nsuid int64, nsgid int64) {
+	if args.fstype == "hugetlbfs" && args.data == "" {
+		args.data = fmt.Sprintf("uid=%d,gid=%d", nsuid, nsgid)
+		args.shift = false
+	}
```
Comment (stgraber) [inline lxd/seccomp/seccomp.go]: This only handles the case where the user didn't specify a uid/gid.

If `uid` or `gid` are specified by the user, we should parse them and shift them using the instance map.

File: lxd/instance/drivers/driver_lxc.go [hunk 3]
```diff
@@ -4291,6 +4291,32 @@ func (c *lxc) Update(args db.InstanceArgs, userRequested bool) error {
 						return err
 					}
 				}
+			} else if strings.HasPrefix(key, "limits.hugepages.") {
```
Comment (stgraber) [inline lxd/instance/drivers/driver_lxc.go]: This handles the `Update` case but doesn't cover the normal startup case.

File: doc/api-extensions.md [hunk 4]
```diff
@@ -926,3 +926,16 @@ Introduces the ability to create a storage pool from an existing non-empty volum
...
+
+##limits\_hugepages.*
```
Comment (stgraber) [inline doc/api-extensions.md]: Missing space and invalid name (*) for an extension.

Comment (stgraber) [reply doc/api-extensions.md]: Should just be `## limits\_hugepages`

File: doc/instances.md [hunk 5]
```diff
@@ -46,6 +46,10 @@ limits.cpu                                  | string    | - (all)           | ye
...
 limits.disk.priority                        | integer   | 5 (medium)        | yes           | -                 | When under load, how much priority to give to the instance's I/O requests (integer between 0 and 10)
+limits.hugepages.64KB                       | string    | -                 | yes           | container         | Fixed value in bytes (various suffixes supported, see below) to limit number of 64 KB hugepages
+limits.hugepages.1MB                        | string    | -                 | yes           | container         | Fixed value in bytes (various suffixes supported, see below) to limit number of 1 MB hugepages
+limits.hugepages.2MB                        | string    | -                 | yes           | container         | Fixed value in bytes (various suffixes supported, see below) to limit number of 2 MB hugepages
+limits.hugepages.1GB                        | string    | -                 | yes           | container         | Fixed value in bytes (various suffixes supported, see below) to limit number of 1 GB hugepages
```
Comment (stgraber) [inline doc/instances.md]: We still need a note somewhere that the availability of the various sizes differs based on architecture.

**Comment by brauner** [comment]
On March 2, 2020 6:40:32 PM GMT+01:00, "Stéphane Graber" <notifications@github.com> wrote:
>stgraber commented on this pull request.
>
>
>
>> +	if args.fstype == "hugetlbfs" && args.data == "" {
>+		args.data = fmt.Sprintf("uid=%d,gid=%d", nsuid, nsgid)
>+		args.shift = false
>+	}
>
>This only handles the case where the user didn't specify a uid/gid.
>
>If `uid` or `gid` are specified by the user, we should parse them and
>shift them using the instance map.

Yeah I wasn't sure about that. Or whether we should not guess. But sure.

**Comment by brauner** [comment]
On March 2, 2020 6:42:22 PM GMT+01:00, "Stéphane Graber" <notifications@github.com> wrote:
>stgraber commented on this pull request.
>
>
>
>> @@ -926,3 +926,16 @@ Introduces the ability to create a storage pool
>from an existing non-empty volum
>This option should be used with care, as LXD can then not guarantee
>that volume name conflicts won't occur
> with non-LXD created volumes in the same volume group.
>This could also potentially lead to LXD deleting a non-LXD volume
>should name conflicts occur.
>+
>+## container\_syscall\_intercept\_hugetlbfs
>+When mount syscall interception is enabled and hugetlbfs is specified
>as an
>+allowed filesystem type LXD will mount a separate hugetlbfs instance
>for the
>+container with the uid and gid mount options set to the container's
>root uid
>+and gid. This ensure that processes in the container can use
>hugepages.
>+
>+##limits\_hugepages.*
>
>Missing space and invalid name (*) for an extension.

Ohz I wasn't too hasty it seems.

**Comment by brauner** [comment]
On March 2, 2020 6:43:22 PM GMT+01:00, "Stéphane Graber" <notifications@github.com> wrote:
>stgraber commented on this pull request.
>
>
>
>> +limits.hugepages.64KB                       | string    | -         
>| yes           | container         | Fixed value in bytes (various
>suffixes supported, see below) to limit number of 64 KB hugepages
>+limits.hugepages.1MB                        | string    | -           
>| yes           | container         | Fixed value in bytes (various
>suffixes supported, see below) to limit number of 1 MB hugepages
>+limits.hugepages.2MB                        | string    | -           
>| yes           | container         | Fixed value in bytes (various
>suffixes supported, see below) to limit number of 2 MB hugepages
>+limits.hugepages.1GB                        | string    | -           
>| yes           | container         | Fixed value in bytes (various
>suffixes supported, see below) to limit number of 1 GB hugepages
>
>We still need a note somewhere that the availability of the various
>sizes differs based on architecture.

Probably good to do it in api_extension.md?

**Comment by stgraber** [comment]
No, the doc about hugepages should probably be a separate paragraph in instances.md.
`api_extensions.md` is really just meant for developers.

File: lxd/seccomp/seccomp.go (hunk 2 above)
Comment (brauner) [reply lxd/seccomp/seccomp.go]: done

Comment (brauner) [reply lxd/seccomp/seccomp.go]: Somehow missed that.

File: doc/api-extensions.md (hunk 4 above)
Comment (brauner) [reply doc/api-extensions.md]: fixed

**Comment by brauner** [comment]
> No, the doc about hugepages should probably be a separate paragraph in instances.md.
> `api_extensions.md` is really just meant for developers.

I added a note both to the table entry for each key but added a whole section on hugepage limits too.

**Comment by stgraber** [comment]
So we're still missing the container startup logic. The code in place will handle a live change to the limit just fine but there's no code to apply the limit on initial startup.

Also, can you split the changes to `lxd/cgroup` to their own commit?

**Comment by brauner** [comment]
> So we're still missing the container startup logic. The code in place will handle a live change to the limit just fine but there's no code to apply the limit on initial startup.
> 
> Also, can you split the changes to `lxd/cgroup` to their own commit?

Yeah, I've missed to finish that part yesterday. Though you might not have caught my message aobut it.

**Pull request merged**